src/net.cpp

```diff
@@ -340,71 +340,161 @@ bool CConnman::CheckIncomingNonce(uint64_t nonce)
+            FindNode(addrConnect.ToStringIPPort()))
+            return false;
+    } else if (FindNode(std::string(pszDest))) {
+        return false;
     }
 
     /// debug print
     LogPrint(BCLog::NET, "trying connection %s lastseen=%.1fhrs\n",
         pszDest ? pszDest : addrConnect.ToString(),
         pszDest ? 0.0 : (double)(GetAdjustedTime() - addrConnect.nTime)/3600.0);
 
-    // Connect
-    SOCKET hSocket;
-    bool proxyConnectionFailed = false;
-    if (pszDest ? ConnectSocketByName(addrConnect, hSocket, pszDest, Params().GetDefaultPort(), nConnectTimeout, &proxyConnectionFailed) :
-                  ConnectSocket(addrConnect, hSocket, nConnectTimeout, &proxyConnectionFailed))
-    {
-        if (!IsSelectableSocket(hSocket)) {
-            LogPrintf("Cannot create connection: non-selectable socket created (fd >= FD_SETSIZE ?)\n");
-            CloseSocket(hSocket);
-            return NULL;
+    std::string strDest;
+    int port = Params().GetDefaultPort();
+    CAddress addrDest;
+
+    // Check input.
+    // If pszDest is set:
+    // - It must contain parseable host, and optionally a port.
+    // - If no port is supplied, the default is used.
+    // Otherwise, addrDest must be valid.
+    if (pszDest) {
+        SplitHostPort(std::string(pszDest), port, strDest);
+        if (strDest.empty()) {
+            return false;
+        }
+    } else {
+        addrDest = addrConnect;
+        if (!addrDest.IsValid()) {
+            return false;
         }
+    }
 
-        if (pszDest && addrConnect.IsValid()) {
-            // It is possible that we already have a connection to the IP/port pszDest resolved to.
-            // In that case, drop the connection that was just created, and return the existing CNode instead.
-            // Also store the name we used to connect in that CNode, so that future FindNode() calls to that
-            // name catch this early.
-            LOCK(cs_vNodes);
-            CNode* pnode = FindNode((CService)addrConnect);
-            if (pnode)
-            {
-                pnode->MaybeSetAddrName(std::string(pszDest));
-                CloseSocket(hSocket);
-                LogPrintf("Failed to open new connection, already connected\n");
-                return NULL;
+    if (!strDest.empty()) {
+        // Resolve strDest, if set. Only actually perform a dns lookup if they're enabled
+        // by the current settings. Note that if strDest is already an ip address, this will
+        // succeed regardless of the lookup setting.
+        std::vector<CService> addrResolved;
+        if (Lookup(strDest.c_str(), addrResolved, port, fNameLookup && !HaveNameProxy(), 256)) {
+            if (addrResolved.size() > 0) {
+                addrDest = CAddress(addrResolved[GetRand(addrResolved.size())], NODE_NONE);
             }
         }
+    }
 
-        addrman.Attempt(addrConnect, fCountFailure);
+    if (!strDest.empty() && addrDest.IsValid()) {
```

Comment:
> By keeping the connection flow separated like this, it's trivial to factor out that way.

Probably there is something going over my head, but I don't understand the reasoning.

Currently this does:

```
if (condition) {
  // bunch of code
} else {
  // other code
}
if (equivalent condition) {
  // code that you want to factor out
}
if (equivalent condition) {
  // more code that you want to factor out
}
```

I don't see why it would be any harder to refactor later with:

```
if (condition) {
  // bunch of code
  // code that you want to factor out
  // more code that you want to factor out
} else {
  // other code
}
```